[Jay Aljelo Saez Ting] iP

jayasting98 opened the pull request:
# Duke CHADbot
> "Refers to a person of the highest standards and that is perfect at everything with no flaws, will act upon what they wish to without hesitation or regard for authority, highly respected by others" -[Urban Dictionary Entry on "Absolute Chad"](https://www.urbandictionary.com/define.php?term=Absolute%20chad)

Duke CHADbot :robot: can _**easily**_ handle your tasks, including
* To-do tasks
* Tasks with deadlines
* Tasks that occur at a particular point in time such as events

Simply do the following to start
1. [Download the JAR file.](https://github.com/jayasting98/ip/releases/tag/v0.1)
1. Run it.
1. Add your tasks.
1. Let Duke CHADbot handle your tasks

If you know programming, it is _**highly extensible**_ too! Simply fork the repo and start modifying the code as you wish. For instance, if you'd like to create more types of commands, simply add a new `CommandType` instance and extend the `Command` class. For example:
```java
public enum CommandType {
    // ...
    YOUR_COMMAND(commandName, commandDescription);
    // ...
}
```
```java
public class YourCommand extends Command {
    // ...

    // define constructor
    public YourCommand(String command) {
        super(command);
        // ...
    }

    // implement abstract methods execute, parseCommand, getCommandType
    @Override
    public void execute(TaskHandler taskHandler, Ui ui) {
        // ...
    }

    @Override
    void parseCommand(String[] tokens) {
        // ...
    }

    @Override
    CommandType getCommandType() {
        // ...
        return CommandType.YOUR_COMMAND;
    }

    // ...
}
```

Features (including planned features):
- [x] Commands
- [x] Tasks
- [ ] GUI (coming soon)
- [ ] Solves halting problem (coming ~~never~~ soon using [this library](https://youtu.be/dQw4w9WgXcQ))

Commit:
Add Gradle support

jayasting98 committed:
Add greeting

jayasting98 committed:
Merge branch 'feature-greet' into develop

jayasting98 committed:
Implement exit command and echo functionality

jayasting98 committed:
Merge branch 'feature-input' into develop

jayasting98 committed:
Merge branch 'release-level-1'

jayasting98 committed:
Implement adding tasks

jayasting98 committed:
Implement listing tasks

jayasting98 committed:
Merge branch 'feature-task' into develop

jayasting98 committed:
Add Javadoc for main class and main method

jayasting98 committed:
Merge branch 'release-level-2' into develop

jayasting98 committed:
Merge branch 'release-level-2'

jayasting98 committed:
Merge branch 'develop' into feature-task

jayasting98 committed:
Implement marking tasks as done

jayasting98 committed:
Merge branch 'feature-task' into develop

jayasting98 committed:
Merge branch 'release-level-3'

jayasting98 committed:
Add tasks of the to-do type, add command parser, and add message formatter

jayasting98 committed:
Add TemporalTask and DeadlineTask classes

jayasting98 committed:
Add EventTask class

jayasting98 committed:
Merge branch 'feature-task' into develop

jayasting98 committed:
Improve encapsulation of MessageFormatter methods

jayasting98 committed:
Update Javadoc

jayasting98 committed:
Merge branch 'release-level-4' into develop

jayasting98 committed:
Merge branch 'release-level-4'

jayasting98 committed:
Merge branch 'develop' into feature-task

jayasting98 committed:
Update test cases

jayasting98 committed:
Merge branch 'feature-task' into develop

jayasting98 committed:
Merge branch 'release-a-text-ui-testing'

jayasting98 committed:
Add DukeInvalidCommandException class

jayasting98 committed:
Merge branch 'feature-task' into develop

jayasting98 committed:
Add Javadoc for the DukeInvalidCommandException class

jayasting98 committed:
Merge branch 'release-level-5' into develop

jayasting98 committed:
Merge branch 'develop' into feature-task

jayasting98 committed:
Merge branch 'release-level-5'

jayasting98 committed:
Implement support for deleting tasks from the list

jayasting98 committed:
Add test cases for deleting tasks

jayasting98 committed:
Format CommandType enum

jayasting98 committed:
Merge branch 'feature-task' into develop

jayasting98 committed:
Format code (minor)

jayasting98 committed:
Merge branch 'release-level-6' into develop

jayasting98 committed:
Merge branch 'release-level-6'

jayasting98 committed:
Implement saving tasks to and loading tasks from storage

jayasting98 committed:
Update tests to delete data before starting tests

jayasting98 committed:
Implement date parsing

jayasting98 committed:
Update tests to use proper date/time format

jayasting98 committed:
Merge branch 'branch-Level-7'

jayasting98 committed:
Merge branch 'branch-Level-8'

jayasting98 committed:
Fix Levels 7 and 8 merge bugs

jayasting98 committed:
Merge branch 'hotfix-level-7-and-8' into develop

jayasting98 committed:
Merge branch 'hotfix-level-7-and-8'

jayasting98 committed:
Extract TaskHandler class

jayasting98 committed:
Extract Ui class

jayasting98 committed:
Extract Command classes

jayasting98 committed:
Clean code, and update Javadoc for all classes

jayasting98 committed:
Update test cases for new Command classes

jayasting98 committed:
Merge branch 'refactor-more-oop' into develop

jayasting98 committed:
Merge branch 'add-gradle-support' into test-unit-introductory

jayasting98 committed:
Integrate Gradle into the project

jayasting98 committed:
Add introductory unit test for the ToDoTask class

jayasting98 committed:
Add introductory unit test for the DeadlineTask class

jayasting98 committed:
Merge branch 'test-unit-introductory' into develop

jayasting98 committed:
Merge branch 'release-more-oop-and-others'

jayasting98 committed:
Implement "Find Tasks" command

jayasting98 committed:
Update test cases for "Find Tasks" command

jayasting98 committed:
Add number of results for "Find Tasks" command

jayasting98 committed:
Update test cases for "Find Tasks" command

jayasting98 committed:
Clean up code to better match coding style

jayasting98 committed:
Add Javadoc for non-trivial, non-private (including non-public) methods

jayasting98 committed:
Merge branch 'branch-A-JavaDoc'

jayasting98 committed:
Merge branch 'branch-A-CodingStandard'

jayasting98 committed:
Merge branch 'branch-Level-9'

jayasting98 committed:
Merge branch 'master' into develop

jayasting98 committed:
Integrate CheckStyle to detect coding style violations

jayasting98 committed:
Fix current minor coding style violations

jayasting98 committed:
Merge branch 'test-checkstyle' into develop

jayasting98 committed:
Add initial JavaFX GUI

jayasting98 committed:
Fix exit command

jayasting98 committed:
Clean code after initial JavaFX GUI trial implementation

jayasting98 committed:
Polish the GUI and give credit for adapted/reused code

jayasting98 committed:
Merge branch 'feature-gui' into develop

jayasting98 committed:
Merge branch 'release-level-10'

andrewbaruch commented on src/main/java/duke/ui/DialogBox.java:
```diff
@@ -0,0 +1,63 @@
+
+/**
+ * Controller class for a single dialog box.
+ * Adapted/reused from https://se-education.org/guides/tutorials/javaFxPart4.html written by Jeffry Lum.
```
Well done on citing your adaptation! :100:

andrewbaruch commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,20 @@
+ */
+public class Duke {
+    /**
+     * Initialises and runs the Duke chatbot.
```
Should this be "Initialize and runs.."? (or Initialise for British English)

jayasting98 replied:
It is intended to be British English, though I guess this module prefers American spelling, so I could just change it. Thanks!

andrewbaruch reviewed (commented):
HI Jay!

I absolutely love your code! It's neat, seems to be thorough at catching the unexpected input, and is definitely easy to look through! I have commented a few thing which may not be in line with the coding standard. Nevertheless, great job!

andrewbaruch commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,20 @@
+import javafx.application.Application;
+
+/**
+ * Main class of Duke
```
Should this have a punctuation after it?

charliemoweng replied:
yes it does look a bit odd.

jayasting98 replied:
Yes, you're right, I missed that. Thanks for helping me spot the mistake!

andrewbaruch commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,202 @@
+
+    private void printFormattedMessage(String message) {
+        String formattedMessage = messageFormatter.getFormattedMessage(message);
+        dialogContainer.getChildren().add(DialogBox.getDukeDialog(formattedMessage, dukeImage));
```
Would this be better if it is broken down to different lines? (to improve readability)

jayasting98 replied:
Yes, I agree. I must have overlooked that. Thank you for pointing it out!

charliemoweng commented on src/main/java/duke/DukeChatbot.java:
```diff
@@ -0,0 +1,82 @@
+import duke.storage.StorageHandler;
+import duke.task.TaskHandler;
+import duke.ui.Ui;
+import javafx.application.Application;
```
maybe can separate them into groups to make it clearer?

jayasting98 replied:
I agree. I normally would separate these if it was my own project and I could set my own coding style. However, CheckStyle considers separating these as an error. I believe it is because separating them is not really a part of CS2103T's Java style guide. So, I will not separate them into groups even though I agree with you and I want to do that.

jayasting98 committed:
Fix code according to pull request reviews

jayasting98 committed:
Merge branch 'fix-pr-reviews' into develop

jayasting98 committed:
Add assertions to check assumptions

Assumptions when methods are called are never actually ensured.

If there are assumptions that are wrong, then they could cause undefined
behaviour in the code, such as errors.

Let's add assertions to check these assumptions to ensure they are true.

These assumptions are based on what a programmer assumes is true.
Assertions are more suitable compared to exceptions for checking errors
caused by the mistakes of a programmer.

jayasting98 committed:
Fix minor bug related to parsing

While the CommandParser can recognise that an input is empty or that a
command does not exist and throw an exception as a result, DukeChatbot
does not handle parsing errors.

If the parsing errors are not handled, then this leads to undefined
behaviour, such as errors like the chatbot not providing feedback to the
user about the fact that empty inputs are not processed.

Let's move the line where the object commandParser parses the input
through its instance method CommandParser::getCommandInstance into the
try block.

Moving the line where the method is called into the try block allows any
exceptions thrown due to parsing invalid inputs to be caught.

jayasting98 committed:
Change one user input error feedback message slightly

jayasting98 committed:
Convert exceptions in Command constructor to assertions

The input to the Command constructor is checked if it corresponds to the
Command sub-class. If it does not, an exception is thrown. This check is
also performed in the CommandParser::getCommandInstance method.

This is unnecessary repetition. Also, if the wrong Command sub-class
constructor is called in the CommandParser::getCommandInstance method,
then it must have been the fault of the programmer, because the check
on the user input was already performed in the CommandParser class.

Let's change the exceptions in the Command class to assertions instead.

Exceptions are for errors caused by the user or other external events
such as I/O errors. These remain handled by the CommandParser class. On
the other hand, assertions are for checking for errors caused by
mistakes by the programmer, usually due to wrong assumptions. Thus, the
exceptions in the Command constructor should be converted to assertions
instead.

jayasting98 committed:
Extract sub-methods in parseCommand for various Command sub-classes

The parseCommand method in the various Command sub-classes tends to be
long. There are many checks and processes that occur within the method
in order to parse the command properly.

Long methods are harder to debug. They also violate software engineering
principles such as SLAP and SRP.

Let's extract various sub-methods from parseCommand for the various
Command sub-classes.

Extracting sub-methods puts a group of lines of code into another
sub-method, replacing those lines with a single line calling that
sub-method, thereby shortening the parseCommand method. The shorter
sub-methods are easier to debug. They also abstract away the
sub-processes done to do checks or to do parsing of the command. The
parseCommand becomes easier to debug, is more readable, and is better at
self-documenting (through the meaningful method names).

jayasting98 committed:
Refactor if block to switch block

jayasting98 committed:
Extract sub-methods from StorageHandler::parseTask

jayasting98 committed:
Extract initialise from DukeChatbot::start

jayasting98 committed:
Merge branch 'develop' into branch-A-CodeQuality

jayasting98 committed:
Merge pull request #1 from jayasting98/branch-A-Assertions

Add increment A-Assertions

jayasting98 committed:
Merge branch 'master' into branch-A-CodeQuality

jayasting98 committed:
Merge pull request #2 from jayasting98/branch-A-CodeQuality

Add increment A-CodeQuality

jayasting98 committed:
Merge branch 'master' into develop

jayasting98 committed:
Add GitHub Actions workflow

jayasting98 committed:
Remove GitHub Actions redirection test steps

jayasting98 committed:
Merge branch 'feature-ci' into develop

jayasting98 committed:
Merge branch 'release-a-ci'

jayasting98 committed:
Add UndoableCommand interface

jayasting98 committed:
Make UndoableCommand package-private

jayasting98 committed:
Add CommandHandler class

jayasting98 committed:
Fix CommandHandler::getInstance

jayasting98 committed:
Add UndoCommand class

jayasting98 committed:
Integrate UndoCommand

jayasting98 committed:
Change method signature of undo to include TaskHandler

jayasting98 committed:
Fix UndoCommand::getCommandType

jayasting98 committed:
Make AddTaskCommand a sub-type of UndoableCommand

jayasting98 committed:
Make ApplyOnTaskNumberCommand a sub-type of UndoableCommand

jayasting98 committed:
Integrate UndoCommand with Ui to allow Duke to respond to undo

jayasting98 committed:
Merge branch 'feature-undo' into develop

jayasting98 committed:
Merge branch 'release-bcd-extension-c-undo'

jayasting98 committed:
Improve GUI

jayasting98 committed:
Rename Duke to Chad in order to match similar GUI rename

jayasting98 committed:
Merge branch 'feature-better-gui' into develop

jayasting98 committed:
Merge branch 'develop'

jayasting98 committed:
Clean and finalise code for v0.2

jayasting98 committed:
Merge branch 'fix-clean-and-finalise' into develop

jayasting98 committed:
Add initial draft of ChadBot User Guide

jayasting98 committed:
Fix coding style errors possibly caused by the renaming of the chatbot

jayasting98 committed:
Fix minor errors in user guide

jayasting98 committed:
Try gif version of Ui.png

jayasting98 committed:
Trim front part of gif version of Ui.png

jayasting98 committed:
Fix bug with resizing that caused undesired scrolling to bottom

jayasting98 committed:
Merge branch 'fix-clean-and-finalise' into develop

jayasting98 committed:
Merge branch 'develop' into misc-user-guide

jayasting98 committed:
Add demonstration gifs and fix minor things in the user guide

jayasting98 committed:
Update GitHub repo README.md to match current product state

jayasting98 committed:
Merge branch 'misc-user-guide' into develop

jayasting98 committed:
Merge branch 'release-v0.2'

jayasting98 committed:
Remove magic literals

jayasting98 committed:
Merge branch 'hotfix-v0.2.1'
